src/server.ts

```diff
@@ -47,128 +39,180 @@ export interface ServerOptions {
   // triggerGenerator?: (name: string, args: Object, context?: Object) => Array<{name: string, filter: Function}>;
 }
 
-interface TriggerAction {
-  name: string;
-  rootValue: any;
-  contextValue?: any;
-}
-
 class Server {
   private onSubscribe: Function;
-  private wsServer: WebSocketServer;
+  private onUnsubscribe: Function;
+  private onConnect: Function;
+  private onDisconnect: Function;
+  private wsServer: WebSocket.Server;
+  private initResult: any;
   private subscriptionManager: SubscriptionManager;
 
-  constructor(options: ServerOptions, httpServer: HttpServer) {
-    const { subscriptionManager, onSubscribe, keepAlive } = options;
+  constructor(options: ServerOptions, socketOptions: WebSocket.IServerOptions) {
+    const {subscriptionManager, onSubscribe, onUnsubscribe, onConnect, onDisconnect, keepAlive} = options;
 
     if (!subscriptionManager) {
       throw new Error('Must provide `subscriptionManager` to websocket server constructor.');
     }
 
     this.subscriptionManager = subscriptionManager;
     this.onSubscribe = onSubscribe;
+    this.onUnsubscribe = onUnsubscribe;
+    this.onConnect = onConnect;
+    this.onDisconnect = onDisconnect;
 
     // init and connect websocket server to http
-    this.wsServer = new WebSocketServer({
-      httpServer: httpServer,
-      autoAcceptConnections: false,
-      // TODO: origin filter
-    });
+    this.wsServer = new WebSocket.Server(socketOptions || {});
+
+    this.wsServer.on('connection', (request: WebSocket) => {
+      request.pause();
+      if (request.protocol === undefined || request.protocol.indexOf(GRAPHQL_SUBSCRIPTIONS) === -1) {
+        request.close(1002);
+        request.terminate();
 
-    this.wsServer.on('request', (request) => {
-      if (request.requestedProtocols.indexOf(GRAPHQL_SUBSCRIPTIONS) === -1) {
-        request.reject(400, 'Unsupported protocol.');
         return;
       }
 
       // accept connection
-      const connection: Connection = request.accept(GRAPHQL_SUBSCRIPTIONS, request.origin);
+      request.resume();
```

Comment:
Huh, we call resume synchronously with pause? That seems weird... why call pause in the first place?  Also pause/resume is the ancient streams API, is this really the right thing to use?